File: adventofcode/2022/day01/day01.rb
# frozen_string_literal: true

def read_input
  ARGF.readlines
end

def part1 sums
  sums.max
end

def part2 sums
  sums.sort.reverse[0...3].sum
end

cals = read_input.map(&:to_i)
cal_sums = cals.chunk_while { |_, n2| n2 != 0 }.map(&:sum)

p part1 cal_sums
p part2 cal_sums

File: adventofcode/2022/day01/day01.hs
import Data.List (sort)

split :: (a -> Bool) -> [a] -> [[a]]
split p xs = f xs []
  where f [] agg = [agg]
        f (y : ys) agg
          | p y = agg : f ys []
          | otherwise = f ys (agg ++ [y])


part1 :: [String] -> Int
part1 = maximum . map (sum . map read) . split null

part2 :: [String] -> Int
part2 = sum . take 3 . reverse . sort . map (sum . map read) . split null

main :: IO ()
main = interact $ unlines . map show . sequence [part1, part2] . lines
